src/main/kotlin/com/cognifide/gradle/aem/instance/Instance.kt

```diff
@@ -62,24 +59,65 @@ interface Instance : Serializable {
-                val nameParts = StringUtils.substringBetween(it.key, ".remote.", ".httpUrl")!!.split("-")
+        fun properties(project: Project): List<Instance> {
+            val localInstances = collectProperties(project, "local").map {
+                val (name, props) = it
+                val nameParts = name.split("-")
                 if (nameParts.size != 2) {
-                    throw InstanceException("Instance list property '${it.key}' does not have valid name part.")
+                    throw InstanceException("Local instance name has invalid format: '$name'.")
                 }
-
                 val (environment, typeName) = nameParts
-                val httpUrl = it.value as String?
-                if (httpUrl.isNullOrBlank()) {
-                    throw InstanceException("Instance list property '${it.key}' value cannot be blank.")
+                val httpUrl = props["httpUrl"]
+                        ?: throw InstanceException("Local instance named '$name' must have property 'httpUrl' defined.")
+
+                LocalInstance.create(httpUrl, {
+                    this.environment = environment
+                    this.typeName = typeName
+
+                    props["password"]?.let { this.password = it }
+                    props["jvmOpts"]?.let { this.jvmOpts = it.split(" ") }
+                    props["startOpts"]?.let { this.startOpts = it.split(" ") }
+                    props["runModes"]?.let { this.runModes = it.split(",") }
+                    props["debugPort"]?.let { this.debugPort = it.toInt() }
+                })
+            }.sortedBy { it.name }
+
+            val remoteInstances = collectProperties(project, "remote").map {
+                val (name, props) = it
+                val nameParts = name.split("-")
+                if (nameParts.size != 2) {
+                    throw InstanceException("Remote instance name has invalid format: '$name'.")
                 }
+                val (environment, typeName) = nameParts
+                val httpUrl = props["httpUrl"]
+                        ?: throw InstanceException("Remote instance named '$name' must have property 'httpUrl' defined.")
 
-                RemoteInstance.create(httpUrl!!, {
+                RemoteInstance.create(httpUrl, {
                     this.environment = environment
                     this.typeName = typeName
+
+                    props["user"]?.let { this.user = it }
+                    props["password"]?.let { this.password = it }
+
                 })
             }.sortedBy { it.name }
+
+            return localInstances + remoteInstances
+        }
+
+        private fun collectProperties(project: Project, type: String): MutableMap<String, MutableMap<String, String>> {
+            return project.properties.filterKeys { Patterns.wildcard(it, "aem.instance.$type.*.*") }.entries.fold(mutableMapOf(), { result, e ->
```

Comment:
Why not Single-Expression functions?

Reply:
even IntelliJ is not enforcing this, because it's more related with own code style.

If expression is not multi-line, ok I used to use single-expression function. If not, not :)